File: dct.py
import math
import random
import time

sz = 256
array2d = []

for x in range(sz):
    array = [random.randint(0, 255) for _ in range(sz)]
    array2d.append(array)


def dct2d(array2d):
    sz = len(array2d)
    coef = dct_coef(sz)
    temp = [[0] * sz for _ in range(sz)]
    result = [[0] * sz for _ in range(sz)]

    for x in range(sz):
        for i in range(sz):
            sum = 0
            for j in range(sz):
                sum += array2d[x][j] * coef[j][i]
            temp[x][i] = sum

    for y in range(sz):
        for i in range(sz):
            sum = 0
            for j in range(sz):
                sum += temp[j][y] * coef[j][i]
            result[y][i] = sum

    return result

def dct_coef(sz):
    fact = math.pi / sz
    coef = [[0] * sz for _ in range(sz)]

    for i in range(sz):
        mult = i * fact
        for j in range(sz):
            coef[j][i] = math.cos((j + 0.5) * mult)

    return coef

start_time = time.time()
result = dct2d(array2d)
print('Time: {:.3f}'.format(time.time() - start_time))

File: dct.js
const sz = 256;
const array2d = [];

for (let x = 0; x < sz; x++) {
  const array = [];
  for (let y = 0; y < sz; y++) {
    array.push(Math.floor(Math.random() * 256));
  }
  array2d.push(array);
}

function dct2d(array2d) {
  const sz = array2d.length;
  const coef = dctCoef(sz);
  const temp = [];
  const result = [];

  for (let x = 0; x < sz; x++) {
    temp[x] = [];
    for (let i = 0; i < sz; i++) {
      let sum = 0;
      for (let j = 0; j < sz; j++) {
        sum += array2d[x][j] * coef[j][i];
      }
      temp[x][i] = sum;
    }
  }

  for (let y = 0; y < sz; y++) {
    result[y] = [];
    for (let i = 0; i < sz; i++) {
      let sum = 0;
      for (let j = 0; j < sz; j++) {
        sum += temp[j][y] * coef[j][i];
      }
      result[y][i] = sum;
    }
  }

  return result;
}

function dctCoef(sz) {
  const fact = Math.PI / sz;
  const coef = [];

  for (let i = 0; i < sz; i++) {
    const mult = i * fact;
    coef[i] = [];
    for (let j = 0; j < sz; j++) {
      coef[i][j] = Math.cos((j + 0.5) * mult);
    }
  }

  return coef;
}

// Calling and timing code
const startTime = Date.now();
const result = dct2d(array2d);
const endTime = Date.now();
const executionTime = endTime - startTime;
console.log('Execution time:', executionTime);

